README.md
# Solr actuators

Readiness endpoint to be used as load balancer check.

Solr is considered ready when it is "almost" ready with tracking alfresco.  
This is computed subjectivelly using the lag reported by the SUMMARY screen: solr is ready when the lag is smaller than MAX_LAG.

The script is available at:

    alfresco/xenit/actuators/readiness

In order to check the health of solr server, the out-of-the-box ping handler can be used.

    alfresco/admin/ping

## Usage

Status code is to be used for the health check: 200 if ready or 503 if not yet ready.

The output of the script offers additional information if being ready as well as information about current tracker status (when parameter info is appended to the query).

## How to run integration tests

    ./gradlew integrationTest



CHANGELOG.md
---
title: Changelog - Solr Actuators

# Alfresco Actuators Changelog
## [0.0.3] - UNRELEASED
	

## [0.0.2] - 2021-04-09

### Fixed
* Fixed pom - no dependencies [#2]
	
[#2] https://github.com/xenit-eu/solr-actuators/pull/2
	
	
## [0.0.1] - 2021-03-29

### Added
* Endpoint /xenit/actuators/readiness [#1]
	
[#1]: https://github.com/xenit-eu/solr-actuators/pull/1


integration-tests/src/test/java/eu/xenit/actuators/integrationtesting/SolrSmokeTests.java
package eu.xenit.actuators.integrationtesting;

import io.restassured.builder.RequestSpecBuilder;
import io.restassured.specification.RequestSpecification;
import org.junit.BeforeClass;
import org.junit.Test;


import static io.restassured.RestAssured.given;
import static java.lang.Thread.sleep;
import static org.hamcrest.core.StringContains.containsString;

public class SolrSmokeTests {
    static RequestSpecification spec;

    @BeforeClass
    public static void setup() {
        String flavor = System.getProperty("flavor");
        String basePathSolr = "solr/alfresco/xenit/actuators/readiness";
	if("solr4".equals(flavor))
	    basePathSolr = "solr4/alfresco/xenit/actuators/readiness";
        String solrHost = System.getProperty("solr.host");

        int solrPort = 0;
        try {
            solrPort = Integer.parseInt(System.getProperty("solr.tcp.8080"));
        } catch(NumberFormatException e) {
            System.out.println("Solr port 8080 is not exposed, probably ssl is enabled");
        }

	System.out.println("Looking at " + solrHost + ":" + solrPort + "/" + basePathSolr + " where flavor=" + flavor);
        String baseURISolr = "http://" + solrHost;

        spec = new RequestSpecBuilder()
                .setBaseUri(baseURISolr)
                .setPort(solrPort)
                .setBasePath(basePathSolr)
                .build();
    }


    @Test
    public void testActuatorsEndpoint() {
        // wait until solr tracks
        long sleepTime = 30000;
        try {
            sleep(sleepTime);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }


        given()
                .spec(spec)
                .when()
                .get()
                .then()
                .statusCode(200)
                .body(containsString("UP"))
                .toString();

    }
}


solr-actuators/src/main/java/eu/xenit/actuators/handler/ReadinessHandler.java
package eu.xenit.actuators.handler;


import org.alfresco.solr.AlfrescoCoreAdminHandler;
import org.alfresco.solr.TrackerState;
import org.alfresco.solr.tracker.AclTracker;
import org.alfresco.solr.tracker.MetadataTracker;
import org.alfresco.solr.tracker.TrackerRegistry;
import org.apache.solr.common.SolrException;
import org.apache.solr.handler.RequestHandlerBase;
import org.apache.solr.request.SolrQueryRequest;
import org.apache.solr.response.SolrQueryResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Method;

public class ReadinessHandler extends RequestHandlerBase {

    private static final Logger logger = LoggerFactory.getLogger(ReadinessHandler.class);
    private static final long MAX_LAG = 1800L;

    Object invokeMethods(Object target, String... methods) throws Exception {
        for (String methodName : methods) {
            Method method = target.getClass().getMethod(methodName);
            target = method.invoke(target);
        }
        return target;
    }


    @Override
    public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throws Exception {
        AlfrescoCoreAdminHandler coreAdminHandler = null;
        if(isSolr6(req)) {
            coreAdminHandler = (AlfrescoCoreAdminHandler) invokeMethods(req,"getCore","getCoreContainer","getMultiCoreHandler");
        } else {
            coreAdminHandler = (AlfrescoCoreAdminHandler) invokeMethods(req,"getCore","getCoreDescriptor","getCoreContainer","getMultiCoreHandler");
        }

        long lastTxCommitTimeOnServer = 0;
        long lastChangeSetCommitTimeOnServer = 0;
        TrackerState metadataTrackerState = null;
        TrackerState aclsTrackerState = null;

        try {
            TrackerRegistry trackerRegistry = coreAdminHandler.getTrackerRegistry();
            String coreName = req.getCore().getName();

            MetadataTracker metadataTracker = trackerRegistry.getTrackerForCore(coreName, MetadataTracker.class);
            AclTracker aclTracker = trackerRegistry.getTrackerForCore(coreName, AclTracker.class);

            metadataTrackerState =  metadataTracker.getTrackerState();
            aclsTrackerState = aclTracker.getTrackerState();
            lastTxCommitTimeOnServer = metadataTrackerState.getLastTxCommitTimeOnServer();
            lastChangeSetCommitTimeOnServer = aclsTrackerState.getLastChangeSetCommitTimeOnServer();
        }  catch (Exception e) {
            rsp.add("ready", "DOWN");
            rsp.setException(new SolrException(SolrException.ErrorCode.SERVICE_UNAVAILABLE, e.getMessage()));
            return;
        }

        if((lastTxCommitTimeOnServer == 0 || lastChangeSetCommitTimeOnServer == 0)) {
            rsp.add("ready","DOWN");
            rsp.setException(new SolrException(SolrException.ErrorCode.SERVICE_UNAVAILABLE,"Solr did not yet get latest values from server"));
            return;
        }

        long lastIndexChangeSetCommitTime = aclsTrackerState.getLastIndexedChangeSetCommitTime();
        long lastIndexTxCommitTime = metadataTrackerState.getLastIndexedTxCommitTime();
        long txLagSeconds = (lastTxCommitTimeOnServer - lastIndexTxCommitTime) / 1000;
        if(req.getParams().get("info")!=null) {
            rsp.add("txLag", txLagSeconds);
            rsp.add("lastTxCommitTimeOnServer", lastTxCommitTimeOnServer);
        }
        long changeSetLagSeconds = (lastChangeSetCommitTimeOnServer - lastIndexChangeSetCommitTime) / 1000;
        if(req.getParams().get("info")!=null) {
            rsp.add("changeSetLag", changeSetLagSeconds);
            rsp.add("lastChangeSetCommitTimeOnServer", lastChangeSetCommitTimeOnServer);
        }

        if(txLagSeconds >= MAX_LAG || changeSetLagSeconds >= MAX_LAG) {
            rsp.add("ready","NO");
            rsp.setException(new SolrException(SolrException.ErrorCode.SERVICE_UNAVAILABLE,"Lag is larger than permitted: txLag=" + txLagSeconds + ", changeSetLag=" + changeSetLagSeconds + ", MAX_LAG=" + MAX_LAG));
            return;
        } else {
            rsp.add("ready","UP");
        }
    }

    private boolean isSolr6(SolrQueryRequest req) {
        try {
            invokeMethods(req,"getCore","getCoreContainer");
        } catch (Exception e) {
            return false;
        }
        return true;
    }

    @Override
    public String getDescription() {
        return null;
    }

    @Override
    public String getSource() {
        return null;
    }
}


integration-tests/src/java/eu/xenit/actuators/handler/SolrSmokeTests.java
package eu.xenit.actuators.handler;

import io.restassured.builder.RequestSpecBuilder;
import io.restassured.specification.RequestSpecification;
import org.junit.BeforeClass;
import org.junit.Test;

import static io.restassured.RestAssured.given;
import static java.lang.Thread.sleep;
import static org.hamcrest.core.StringContains.containsString;

public class SolrSmokeTests {
    static RequestSpecification spec;

    @BeforeClass
    public static void setup() {

        String basePathSolr = "solr/alfresco/xenit/actuators/readiness";
        String solrHost = System.getProperty("solr.host");

        int solrPort = 0;
        try {
            solrPort = Integer.parseInt(System.getProperty("solr.tcp.8080"));
        } catch(NumberFormatException e) {
            System.out.println("Solr port 8080 is not exposed, probably ssl is enabled");
        }

        String baseURISolr = "http://" + solrHost;

        spec = new RequestSpecBuilder()
                .setBaseUri(baseURISolr)
                .setPort(solrPort)
                .setBasePath(basePathSolr)
                .build();
    }


    @Test
    public void testActuatorsEndpoint() {
        // wait until solr tracks
        long sleepTime = 30000;
        try {
            sleep(sleepTime);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }


        given()
                .spec(spec)
                .when()
                .get()
                .then()
                .statusCode(200)
                .body(containsString("UP"))
                .toString();

    }
}
